ledger/src/blockstore.rs

```diff
@@ -2379,8 +2379,15 @@ impl Blockstore {
             "blockstore-rpc-api",
             ("method", "get_transaction_status", String)
         );
-        self.get_transaction_status_with_counter(signature, confirmed_unrooted_slots)
-            .map(|(status, _)| status)
+        let (status, _) =
+            self.get_transaction_status_with_counter(signature, confirmed_unrooted_slots)?;
+        if let Some((slot, _)) = status {
+            let first_available_block = self.get_first_available_block()?;
+            if slot < first_available_block {
+                return Ok(None);
+            }
```

Comment:
I'm wondering if we need this check here after the fact. Within `get_transaction_status_with_counter()`, we get the `lowest_cleanup_slot` and build our iterator in the forward direction from that slot
https://github.com/solana-labs/solana/blob/64b36135cba574cb6c3b5d0404b52da2c6cd9c6c/ledger/src/blockstore.rs#L2331-L2337

So, it is the case that if we find a matching sig, then the slot the sig was in is `>= lowest_cleanup_slot`. We then grab the transaction status and return it; assuming it is valid, I see no reason to discard it at the end.

Also, `get_transaction_status_with_counter()` holds the `lowest_cleanup_slot` lock. Imagine the following sequence:
1) `get_transaction_status()` calls `get_transaction_status_with_counter()`
2) `get_transaction_status_with_counter()` grabs the `lowest_cleanup_slot` lock, which is at slot `X`
3) `get_transaction_status_with_counter()` finds the transaction status for desired sig in slot `X + i`
4) `get_transaction_status_with_counter()` returns and gives up `lowest_cleanup_slot` lock
5) `LedgerCleanupService` comes along and does some cleanup, advancing `lowest_cleanup_slot` to `X + c` where `c > i`
6) get_transaction_status()` checks the result slot against `get_first_available_block()`; it is the case that `get_first_available_block() > X + c`
7) From 5), `i < c` so the result will get discarded and we'll return `Ok(None)`

This isn't a dangerous race condition necessarily (and given the timing required, I think this would be incredibly unlikely), but again, I don't see any reason to discard the result if we already looked it up and found it

Reply:
Good point, maybe hoist this logic to `get_transaction_status_with_counter()`, inside the lock, then.
If we don't, the `Some()` cases in the before and until blocks in `get_confirmed_signatures_for_address2()` could return a slot/lowest_slot > first_available_block, because lowest_cleanup_slot is not necessarily (in fact usually isn't) the same as first_available_block.

Reply:
Technically, I think we already have the race you described between `get_transaction_status()` and `get_confirmed_signatures_for_address2()`, because we pull the whole block after releasing the `lowest_cleanup_slot` lock. But fixing that, if we choose to, can be independent of this change.

Reply:
> because lowest_cleanup_slot is not necessarily (in fact usually isn't) the same as first_available_block.

Talking this through "out-loud" to help gather my thoughts ... here is the definition of `get_first_available_block()`:
https://github.com/solana-labs/solana/blob/ecb1f8a9d739213cca027660d74b3b090c0d5d3e/ledger/src/blockstore.rs#L2002-L2013
And here is `lowest_slot_with_genesis()`:
https://github.com/solana-labs/solana/blob/ecb1f8a9d739213cca027660d74b3b090c0d5d3e/ledger/src/blockstore.rs#L3385-L3396

So, in normal conditions, `lowest_slot_with_genesis()` will return the first slot with any shreds (not necessarily complete and thus also not necessarily a root). Let's call this slot `S_l`. `get_first_available_block()` then looks for the first rooted slot from `S_l`, and then returns the next rooted slot. So, if `S_l` did happen to be a root, the result of `get_first_available_block()` will be `S_l`'s child

`lowest_cleanup_slot` is the most recent slot that we purged. So, assuming `lowest_cleanup_slot == C`, then
- First root in database `>= C + 1`
- `get_first_available_block()` result `>= C + 2`